File: test/features/achievements/achievement_model_cache_test.dart
import 'dart:convert';

import 'package:clashkingapp/features/achievements/data/achievement_model_cache.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';

void main() {
  test(
    'downloads each model once and reuses its in-memory data source',
    () async {
      var requests = 0;
      final cache = AchievementModelCache(
        client: MockClient((request) async {
          requests++;
          return http.Response.bytes(<int>[1, 2, 3, 4], 200);
        }),
      );

      final sources = await Future.wait([
        cache.resolve('https://assets.example/badge.glb'),
        cache.resolve('https://assets.example/badge.glb'),
      ]);
      final reopened = await cache.resolve('https://assets.example/badge.glb');

      expect(requests, 1);
      expect(sources[0], sources[1]);
      expect(reopened, sources[0]);
      expect(
        sources[0],
        'data:model/gltf-binary;base64,${base64Encode(<int>[1, 2, 3, 4])}',
      );
      expect(cache.peek('https://assets.example/badge.glb'), sources[0]);
    },
  );

  test('falls back to the remote URL when preloading fails', () async {
    final cache = AchievementModelCache(
      client: MockClient((request) async => http.Response('', 503)),
    );
    const url = 'https://assets.example/badge.glb';

    expect(await cache.resolve(url), url);
  });

  test('falls back to the remote URL when the download times out', () async {
    final cache = AchievementModelCache(
      timeout: const Duration(milliseconds: 1),
      client: MockClient((request) async {
        await Future<void>.delayed(const Duration(milliseconds: 20));
        return http.Response('', 200);
      }),
    );

    expect(
      await cache.resolve('https://assets.example/slow-badge.glb'),
      'https://assets.example/slow-badge.glb',
    );
  });
}

File: expo/src/features/achievements/data/achievement-model-cache.test.ts
import { AchievementModelCache } from './achievement-model-cache';

function binaryResponse(bytes: readonly number[], status = 200): Response {
  return {
    ok: status >= 200 && status < 300,
    status,
    arrayBuffer: async () => Uint8Array.from(bytes).buffer,
  } as Response;
}

test('downloads a model once and reuses the in-memory data source', async () => {
  const fetchImplementation = jest.fn(async () => binaryResponse([1, 2, 3, 4]));
  const cache = new AchievementModelCache({
    fetchImplementation: fetchImplementation as typeof fetch,
  });
  const url = 'https://assets.example/badge.glb';
  const [first, second] = await Promise.all([cache.resolve(url), cache.resolve(url)]);
  expect(first).toBe('data:model/gltf-binary;base64,AQIDBA==');
  expect(second).toBe(first);
  expect(await cache.resolve(url)).toBe(first);
  expect(cache.peek(url)).toBe(first);
  expect(fetchImplementation).toHaveBeenCalledTimes(1);
});

test('memoizes the remote URL when the preload fails', async () => {
  const fetchImplementation = jest.fn(async () => binaryResponse([], 503));
  const cache = new AchievementModelCache({
    fetchImplementation: fetchImplementation as typeof fetch,
  });
  const url = 'https://assets.example/badge.glb';
  expect(await cache.resolve(url)).toBe(url);
  expect(await cache.resolve(url)).toBe(url);
  expect(fetchImplementation).toHaveBeenCalledTimes(1);
});

test('aborts a timed-out preload and falls back to the remote URL', async () => {
  const fetchImplementation = jest.fn(
    (_url: RequestInfo | URL, init?: RequestInit) =>
      new Promise<Response>((_resolve, reject) => {
        init?.signal?.addEventListener('abort', () => reject(new Error('aborted')));
      }),
  );
  const cache = new AchievementModelCache({
    fetchImplementation: fetchImplementation as typeof fetch,
    timeoutMs: 1,
  });
  const url = 'https://assets.example/slow.glb';
  expect(await cache.resolve(url)).toBe(url);
  expect(cache.peek(url)).toBe(url);
});
